test_plans/oneapi/default_context.asciidoc

```diff
@@ -0,0 +1,84 @@
+:sectnums:
+:xrefstyle: short
+
+= Test plan for sycl_ext_oneapi_default_context
+
+This is a test plan for the default SYCL context extension defined in
+https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_default_context.asciidoc[sycl_ext_oneapi_default_context.asciidoc].
+
+== Testing scope
+
+=== Device coverage
+
+All of the tests described below are performed only on the default device that
+is selected on the CTS command line.
+
+=== Feature test macro
+
+All of the tests should use `#ifdef SYCL_EXT_ONEAPI_DEFAULT_CONTEXT` so they can
+be skipped if feature is not supported.
+
+== Tests
+The extension defines a default context that can be retrieved using the following new method for the platform class:
+
+[source,c++]
+----
+context ext_oneapi_get_default_context()
+----
+
+It also updates the `queue` class to use this context instead of creating a new one.
+
+=== Default Context Devices Test
+
+The default context should contain all of a platform's device. The test should create a platform and get its default context. The test should then check that the context has all of the devices from the platform.
+
+=== Queue Constructor Test
+
+When constructing a `queue` without specifying a `context`, the `queue` should use the default context. The test should create a new `context` using the default constructor and retrieve the platform's default context. The test should then do the following:
```

Comment:
> I didn't realize that this was valid. I think I can change this test to construct a new context using only the default device or I can remove the check that the new context is not equal to the default context.

Actually, I take this back.  I think it would not be valid for the default-constructed `context` to be the same object as the platform's default context.

Reply:
I think it comes from this text in [Section 4.5.2. Common reference semantics](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:reference-semantics):

> T must be equality comparable in the host application. Equality between two instances of T (i.e. a == b) must be true if one instance is a copy of the other and **non-equality between two instances of T (i.e. a != b) must be true if neither instance is a copy of the other**

(Emphasis mine.)

In the case we're talking about, the application creates a context with the context's default constructor.  This is not the copy constructor, so it is not making a copy.  Therefore, the spec statement above requires this context object to compare unequal with other context objects.

Reply:
I am puzzled. This is probably to be clarified in the spec (or in my brain if the spec is clear :smile:).
Creating a default queue gives a different queue every time because this is the goal to express concurrency.
Creating a default device will probably give the same device every time, specially when there is only 1 device.
So what about the context? Is there a value to have a different context every time? :thinking:

Reply:
This could be a topic to discuss in the WG.  My comment above merely describes the current state of the specification.  If the WG wants to allow the context constructors to return copies of existing `context` objects, I think they need to change the specification.

I suppose there might be an advantage to constructing separate `context` objects.  We use `context` as a sort of container for USM memory allocations.  USM memory allocated from context _A_ can only be accessed from a kernel submitted to a queue that also uses context _A_.  If an application has USM memory buffer _B1_ that is only used in kernel _K1_ and USM memory buffer _B2_ that is only used in kernel _K2_, there might be an advantage to using two different contexts.  Perhaps the implementation could perform some sort of optimization when it sees the separate contexts?